ping3.py

```diff
@@ -68,14 +74,26 @@ def receive_one_ping(my_socket, ID, timeout):
         # Filters out the echo request itself.
         # This can be tested by pinging 127.0.0.1
         # You'll see your own request
-        if type != 8 and packetID == ID:
-            bytesInDouble = struct.calcsize("d")
-            timeSent = struct.unpack("d", recPacket[28:28 + bytesInDouble])[0]
-            return timeReceived - timeSent
-
-        timeLeft = timeLeft - howLongInSelect
-        if timeLeft <= 0:
-            return
+        if type != icmp.ECHO_REQUEST and packetID == ID:
+            if type == icmp.ECHO_REPLY:
+                bytesInDouble = struct.calcsize("d")
+                timeSent = struct.unpack("d", recPacket[28:28 + bytesInDouble])[0]
+                return timeReceived - timeSent
+            elif type == icmp.TIME_EXCEEDED:
```

Comment:
I cannot trigger `icmp.TIME_EXCEEDED`, because when `type == icmp.TIME_EXCEEDED` the packetID is 0, and then the TimeoutException raised. Did not figure out why.